### ISSUE TITLE
Automate version updating (goodly)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2885]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#2885|
|Type|enhancement|
|Created|2007-11-12 19:59:20Z|
|Branch|https://github.com/twisted/twisted/tree/change-versions-2885-2|

There should be some first-class code for updating version numbers in a Twisted branch.

<details><summary>Searchable metadata</summary>

```
trac-id__2885 2885
type__enhancement enhancement
reporter__radix radix
priority__highest highest
milestone__twisted_8_0 twisted-8.0
branch__branches_change_versions_2885_2 branches/change-versions-2885-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1194897560000000 1194897560000000
changetime__1195592527000000 1195592527000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

(In [21717]) Branching to 'change-versions-2885'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

(In [21731]) Branching to 'change-versions-2885-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @jml|
|-|-|

  * twisted/python/test/test_dist.py
    * no copyright or module docstring
    * ```GetVersionTest``` has no class docstring
  * twisted/python/_release.py
    * only the tests use ```getCurrentVersion``` - delete it?

Ran out of time on this one, didn't get to look at most of the changes.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml set owner to @exarkun|
|-|-|

Updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @jml<br>@exarkun set status to `new`|
|-|-|

It'd be good if all the test methods had docstrings, even the ones which just got moved and not modified (at least in test_release, although it'd be nice to spruce up test_dist as well), since supposedly a goal here is "first class release code".

test_findTwistedStyleProjects and test_getProjectName and test_updateTwistedVersionInformation use _twisted_ and _twisted.conch_ even though they're not operating on the real Twisted source.  It might be clearer to use made up names here.  test_updateTwistedVersionInformation also doesn't cover the case where ```baseDirectory``` is ```None```.  There aren't any users of this API either so I can't tell if that is the normal case or not.  Maybe those arguments just shouldn't be optional?

```replaceProjectVersion``` writes out a string which refers to ```admin/change-versions```, which no longer exists.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set owner to @exarkun|
|-|-|


 * Docstrings added for all `test_release` tests. Not so concerned about `test_dist`, since they will likely be heavily modified by the tarball work.
 * `replaceProjectVersion` string changed to no longer refer to `admin/change-versions`.
 * Fixed `test_findTwistedStyleProjects`, `test_getProjectName` and `test_updateTwistedVersionInformation`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jml|
|-|-|

  * twisted.python.test.test_release.ProjectTest.test_updateTwistedVersionInformation failing
  * do something about the XXX in ```twisted.python._release.getNextVersion```
  * updateTwistedVersionInformation still has a couple uncovered lines - the ones that handle the defaults for each parameter
  * ```Project``` could document its two instance attributes.

Otherwise good.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set owner to @exarkun|
|-|-|

It's worth taking another look at this. You too, radix.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @radix|
|-|-|

Looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix set owner to @jml|
|-|-|

I watched commits in my inbox and agree with them. Given that exarkun properly reviewed this, I think it's ready for merging.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml set status to `closed`|
|-|-|

(In [21908]) Unit-tested, understandable API for updating the version of Twisted.

 * Authors: jml, radix
 * Reviewers: exarkun, radix
 * Fixes [#2885](https://github.com/twisted/twisted/issues/2885)

This is part of the new suite of automated release tools. It provides APIs
for updating the version of Twisted and Twisted sub-projects using a new
date-based version system.

Note that it removes the old admin/change-versions script without providing 
a replacement.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `reopened`|
|-|-|

Reverted in [21909]

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_17"></a>|@jml set status to `closed`|
|-|-|

(In [21911]) Unit-tested, understandable API for updating the version of Twisted.

 * Authors: jml, radix
 * Reviewers: exarkun, radix
 * Fixes [#2885](https://github.com/twisted/twisted/issues/2885)

This is part of the new suite of automated release tools. It provides APIs
for updating the version of Twisted and Twisted sub-projects using a new
date-based version system.

Note that it removes the old admin/change-versions script without providing
a replacement.

This re-merge fixes Python 2.3 support.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
